KEYCLOAK-17215 Slowness issue while hitting /auth/admin/realms/$REALM/clients?viewableOnly=true after DELETE a role

rmartinc opened the pull request:
Fix for: https://issues.redhat.com/browse/KEYCLOAK-17215

The idea is avoiding the eviction of all the clients in the realms when a role is deleted. I have tested this and I think it is OK:

1. All the related clients should be correctly evicted from the cache as it is done inside the [HasRolePredicate](https://github.com/keycloak/keycloak/blob/12.0.3/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java#L219). 
2. The delete is done at database level using a query as it's done now for composite roles.
3. The SCOPE_MAPPING table has an [index](https://github.com/keycloak/keycloak/blob/12.0.3/model/jpa/src/main/resources/META-INF/jpa-changelog-3.2.0.xml#L210) on the ROLE_ID column, so this should be fast.

This way only the returned clients of the predicate are evicted and the subsequent search just reloads the ones that really had the role in the scope and were evicted.

rmartinc committed:
KEYCLOAK-17215 Slowness issue while hitting /auth/admin/realms/$REALM/clients?viewableOnly=true after DELETE a role

pskopek commented:
PR moved to community issue: https://issues.redhat.com/browse/KEYCLOAK-17454

pskopek commented:
@hmlnarik can you review this PR, please?

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:
```diff
@@ -338,7 +338,8 @@ public boolean removeRole(RoleModel role) {
         String compositeRoleTable = JpaUtils.getTableNameForNativeQuery("COMPOSITE_ROLE", em);
         em.createNativeQuery("delete from " + compositeRoleTable + " where CHILD_ROLE = :role").setParameter("role", roleEntity).executeUpdate();
-        realm.getClientsStream().forEach(c -> c.deleteScopeMapping(role));
+        String clientScopeMapping = JpaUtils.getTableNameForNativeQuery("SCOPE_MAPPING", em);
```
The method `removeRole` is part of the `RoleProvider` and as such cannot assume that the client scope mappings would be stored in the same JPA store - this holds for the legacy JPA store but will not necessarily be true for new storage.

The right approach would be to introduce a method to delete a client scope mappings based on the role. @vramik, can you please help here?

hmlnarik replied:
Actually, looking a bit further into the code, this should be introduced as a method `preRemove` in a `ClientProvider` that would take a parameter `realm` and `role`, similarly to what's done on L 344 for groups.

rmartinc replied:
Thanks @hmlnarik! I think that I've understood what you mean. I'll add the `preRemove` to the `ClientProvider` method and update the necessary implementations. I'll try to submit a new PR next Monday.

vramik replied:
Thank you @rmartinc for the PR.

+1 for introduction of `preRemove` method.

rmartinc replied:
OK, it's done. There are two things to comment:

1. The method cannot be named `preRemove` because there are two classes that implements Group and Client provider and if the methods have the same signature, the same method is executed several times (x4, x2 in `JpaRealmProvider` and another x2 in `RealmCacheSession`). So I have used the name `preRemoveClientProvider` for the client provider.
2. In the `MapClientProvider` no new searchable field was created, all the clients in the realm are iterated to update them. I think this part is WIP so I prefer to modify it as little as possible.

Just let me know if you want more modifications. Thanks!

hmlnarik commented on server-spi/src/main/java/org/keycloak/models/ClientProvider.java:
```diff
@@ -146,4 +146,12 @@ default ClientModel addClient(RealmModel realm, String clientId) {
+     * @param realm Realm.
+     * @param role Role which will be removed.
+     */
+    void preRemoveClientProvider(RealmModel realm, RoleModel role);
```
Can this be plain `preRemove` similarly to other `preRemove` methods?

rmartinc replied:
@hmlnarik If same name `preRemove`is used both legacy implementations `JpaRealmProvider` and `RealmCacheSession`, the same method should be used for all the implementations, so I needed to include the code for clients in the same one used for groups. I detected that the method was called several times because (x4, x2 becase it was called twice [here](https://github.com/keycloak/keycloak/pull/7815/files#diff-30940e65a4a38797c21073cfeed4d0e4beb58042ed778364f30468571d55a50dR335-R336), and another x2 when the same is done in `RealmCacheSession`). So I decided to rename the method to other signature.

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:
```diff
@@ -587,6 +587,12 @@ public void preRemove(RealmModel realm, RoleModel role) {
+    @Override
+    public void preRemoveClientProvider(RealmModel realm, RoleModel role) {
+        String clientScopeMapping = JpaUtils.getTableNameForNativeQuery("SCOPE_MAPPING", em);
+        em.createNativeQuery("delete from " + clientScopeMapping + " where ROLE_ID = :role").setParameter("role", role.getId()).executeUpdate();
```
Could this be a JPA named query?

rmartinc replied:
I thikn it's not possible using a delete. The reason is that this table is a [collection table](https://github.com/keycloak/keycloak/blob/12.0.4/model/jpa/src/main/java/org/keycloak/models/jpa/entities/ClientEntity.java#L161-L164). It's the same that happens [with CHILD_ROLE a few lines above](https://github.com/keycloak/keycloak/blob/12.0.4/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java#L350). I can use a named query to obtain the list of clients that have the scope mapping (that is done using an inner join) and perform the deletion (`deleteScopeMapping` method) one by one. I decided to do it this way because it is much better (one query instead of N+1) and because it was already done with the CHILD_ROLE. But if you prefer this I can change to that approach.

hmlnarik reviewed (changes requested):
Thank you @rmartinc!

> 1. The method cannot be named `preRemove` because there are two classes that implements Group and Client provider and if the methods have the same signature, the same method is executed several times (x4, x2 in `JpaRealmProvider` and another x2 in `RealmCacheSession`). So I have used the name `preRemoveClientProvider` for the client provider.

You are right. Yet I still believe it should be a single name as the method does the same thing.
Let me take a step back, and I am sorry I missed this obvious solution in my previous review. This is exactly where `ProviderEvent`s fit in. Similarly to how realm removal is broadcast in `removeRealm` method, a `RoleModel.RoleRemoved` provider event should be triggered here, instead of the calls to both `session.groups().preRemove` and `session.clients().preRemove`. That event would be handled in the handler registered in `JpaRealmProviderFactory` which in turn would handle it by calling [`create`](https://github.com/keycloak/keycloak/blob/e126969f8277713d43d560205e473e6142ca91c5/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProviderFactory.java#L50)`(event.session).preRemove(event.getRealm(), event.getRole());`. Similar handler would be registered in `MapClientProviderFactory`.

> 2. In the `MapClientProvider` no new searchable field was created, all the clients in the realm are iterated to update them. I think this part is WIP so I prefer to modify it as little as possible.

This is already stable part of the new storage. Please introduce a new [`SearchableField` into `ClientModel`](https://github.com/keycloak/keycloak/blob/master/server-spi/src/main/java/org/keycloak/models/ClientModel.java#L40-L44) to search by scope, similarly to what has been done for e.g. `UserModel.SearchableFields.ASSIGNED_ROLE` for users in [`MapFieldPredicates`](https://github.com/keycloak/keycloak/blob/e126969f8277713d43d560205e473e6142ca91c5/model/map/src/main/java/org/keycloak/models/map/storage/MapFieldPredicates.java#L91).

rmartinc commented:
OK @hmlnarik, I have update with your new proposal:

* The ClientProvider interface has the method `preRemove` as the other providers.
* The `JpaRealmProvider` now uses the new idea of listening for role removal events and calling its own `preRemove` method at that time. I have checked the method is just executed once, so I think that it's working OK.
* Inside the `MapClientProvider` implementation I have done exactly the same steps that are now done for the `MapGroupProvider` but using `SCOPE_MAPPING_ROLE` instead of `ASSIGNED_ROLE` field.

Let's see how the tests go. As always, if you see something more just let me know.

rmartinc commented:
Thanks a lot @hmlnarik! I see that the idea now is using events everywhere and removing the `preRemove`. I added your commit on top of my changes and merge it into one  single commit. I also adjusted some files to not modify them if unneeded. Tests pass and I have tested locally that the initial problem is also fixed (only the relevant clients are evicted from cache). Let's see the tests....

hmlnarik reviewed (approved):
Thanks @rmartinc! Looks good to me to fix the problem with slow role deletion
